File: python/src/device.py
import operator


class Device:
    def __init__(self, location):
        self.location = location

    def __str__(self):
        return f"{self.location}"

    def best_station(self, stations):
        """ finds the best station for the device based on the maximum power output of the link """
        best_station = max(stations, key=operator.methodcaller(
            'power', location=self.location))
        power = best_station.power(self.location)
        if power == 0:
            return None, power
        else:
            return best_station, power

File: node/src/device.js
class Device {
  constructor(location) {
    this.location = location;
  }

  /**
   * finds the best station for this device based on the largest power output of the link
   * @param stations
   * @returns {Station} best station based on the power output, if nothing was found or device is out of reach {undefined} will be returned
   */
  findBestStation(stations) {
    return stations
      .map(station => {
        return {
          location: station.location,
          reach: station.reach,
          power: station.getPower(station.getDistance(this.location))
        };
      })
      .filter(station => {
        return station.power > 0;
      })
      .sort((stationA, stationB) => {
        return stationA.power > stationB.power;
      })
      .pop();
  }
}

module.exports = Device;
